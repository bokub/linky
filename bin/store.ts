import Conf from 'conf';
import pkg from '../package.json' assert { type: 'json' };

const TOKEN = 'token';

const store = new Conf({
  projectName: 'linky',
  projectVersion: pkg.version,
  migrations: {
    '>=2.0.0': (store: any) => {
      store.clear();
    },
  },
});

export function getToken() {
  return store.get(TOKEN, undefined) as string | undefined;
}

export function saveToken(token: string) {
  return store.set(TOKEN, token);
}
