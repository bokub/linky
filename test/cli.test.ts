import { node } from 'execa';
import jwt from 'jsonwebtoken';
import { getToken } from '../bin/store.js';
import dotenv from 'dotenv';

dotenv.config({ path: '.env' });
process.env.FORCE_COLOR = '0';
const CONSUMPTION_TOKEN = process.env.CONSUMPTION_TOKEN;
const PRODUCTION_TOKEN = process.env.PRODUCTION_TOKEN;

const bin = './bin/cli.ts';
const linky = (args: string) =>
  node(bin, args.split(' '), { nodeOptions: ['--experimental-vm-modules', '--loader=ts-node/esm'] });

describe('cli', () => {
  describe('auth', () => {
    it('should return an error when the token parameter is missing', async () => {
      expect.assertions(1);
      try {
        await linky('auth');
      } catch (e) {
        expect((e as Error).message).toMatch("L'authentification nécessite un token");
      }
    });

    it('should return an error when the token is invalid', async () => {
      expect.assertions(1);
      try {
        await linky('auth -t invalid-token');
      } catch (e) {
        expect((e as Error).message).toMatch('Le token est invalide');
      }
    });

    it('should return an error when the token sub is not an array', async () => {
      expect.assertions(1);
      try {
        await linky(`auth -t ${jwt.sign({ sub: '123' }, 'secret')}`);
      } catch (e) {
        expect((e as Error).message).toMatch('Le token est invalide');
      }
    });

    it('should return an error when the token contains no PRM', async () => {
      expect.assertions(1);
      try {
        await linky(`auth -t ${jwt.sign({ sub: [] }, 'secret')}`);
      } catch (e) {
        expect((e as Error).message).toMatch('Le token est invalide');
      }
    });

    it('should save a token in the store', async () => {
      const token = jwt.sign({ sub: ['1'] }, 'secret');
      const { stdout } = await linky(`auth -t ${token}`);
      expect(getToken()).toEqual(token);
      expect(stdout).toMatch('Votre token a été sauvegardé avec succès');
    });
  });

  describe('daily command', () => {
    it('should display API errors', async () => {
      expect.assertions(1);
      await linky(`auth -t ${CONSUMPTION_TOKEN}`);
      try {
        await linky(`daily -s 2023-04-02 -e 2023-04-01`);
      } catch (e) {
        expect((e as Error).message).toMatch('Start date should be before end date');
      }
    });

    it('should bypass the saved token when there is a token parameter', async () => {
      expect.assertions(2);
      await linky(`auth -t ${jwt.sign({ sub: ['12345123451234'] }, 'secret')}`);
      try {
        await linky(`daily -s 2023-04-01 -e 2023-04-02`);
      } catch (e) {
        expect((e as Error).message).toMatch('Votre token est invalide');
      }

      await expect(linky(`daily -s 2023-04-01 -e 2023-04-02 -t ${CONSUMPTION_TOKEN}`)).resolves.toBeTruthy();
    });

    it('should throw an error when the PRM cannot be accessed', async () => {
      expect.assertions(1);
      await linky(`auth -t ${jwt.sign({ sub: ['12345123451234'] }, 'secret')}`);
      try {
        await linky(`daily -s 2023-04-01 -e 2023-04-02 -p 88888999990000`);
      } catch (e) {
        expect((e as Error).message).toMatch("Ce token ne permet pas d'accéder au PRM");
      }
    });

    it('should use the first of multiple PRMs', async () => {
      expect.assertions(1);
      await linky(`auth -t ${jwt.sign({ sub: ['12345123451234', '1', '123'] }, 'secret')}`);
      try {
        await linky(`daily -s 2023-04-01 -e 2023-04-02`);
      } catch (e) {
        expect((e as Error).message).toMatch('"status": 401');
      }
    });

    it('should use another PRM is there is a prm parameter', async () => {
      expect.assertions(1);
      await linky(`auth -t ${jwt.sign({ sub: ['12345123451234', '1', '123'] }, 'secret')}`);
      try {
        await linky(`daily -s 2023-04-01 -e 2023-04-02 --prm 123`);
      } catch (e) {
        expect((e as Error).message).toMatch('"status": 400');
      }
    });

    // TODO test output, format, quiet
  });

  describe('consumption command', () => {
    beforeAll(async () => {
      await linky(`auth -t ${CONSUMPTION_TOKEN}`);
    });
    test.each(['daily', 'loadcurve', 'maxpower'])('"%s" should display a graph', async (command) => {
      const { stdout } = await linky(`${command} -s 2023-04-01 -e 2023-04-02`);
      expect(stdout).toMatchSnapshot();
    });
  });

  describe('production command', () => {
    beforeAll(async () => {
      await linky(`auth -t ${PRODUCTION_TOKEN}`);
    });
    test.each(['dailyprod', 'loadcurveprod'])('"%s" should display a graph', async (command) => {
      const { stdout } = await linky(`${command} -s 2023-04-01 -e 2023-04-02`);
      expect(stdout).toMatchSnapshot();
    });
  });
});
