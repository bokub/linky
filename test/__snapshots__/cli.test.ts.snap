// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli consumption command "daily" should display a graph 1`] = `
"
Date        Valeur (Wh) Graphique                     
2023-04-01  11776       ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■"
`;

exports[`cli consumption command "loadcurve" should display a graph 1`] = `
"
Date                 Valeur (W) Graphique                     
2023-04-01 00:30:00  752        ■■■■■■■■
2023-04-01 01:00:00  346        ■■■■
2023-04-01 01:30:00  250        ■■■
2023-04-01 02:00:00  154        ■■
2023-04-01 02:30:00  126        ■■
2023-04-01 03:00:00  110        ■■
2023-04-01 03:30:00  140        ■■
2023-04-01 04:00:00  236        ■■■
2023-04-01 04:30:00  138        ■■
2023-04-01 05:00:00  104        ■
2023-04-01 05:30:00  128        ■■
2023-04-01 06:00:00  234        ■■■
2023-04-01 06:30:00  124        ■■
2023-04-01 07:00:00  104        ■
2023-04-01 07:30:00  360        ■■■■
2023-04-01 08:00:00  124        ■■
2023-04-01 08:30:00  122        ■■
2023-04-01 09:00:00  108        ■■
2023-04-01 09:30:00  120        ■■
2023-04-01 10:00:00  106        ■■
2023-04-01 10:30:00  250        ■■■
2023-04-01 11:00:00  108        ■■
2023-04-01 11:30:00  230        ■■■
2023-04-01 12:00:00  1594       ■■■■■■■■■■■■■■■■
2023-04-01 12:30:00  3150       ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
2023-04-01 13:00:00  750        ■■■■■■■■
2023-04-01 13:30:00  1178       ■■■■■■■■■■■■
2023-04-01 14:00:00  566        ■■■■■■
2023-04-01 14:30:00  206        ■■
2023-04-01 15:00:00  440        ■■■■■
2023-04-01 15:30:00  1510       ■■■■■■■■■■■■■■■
2023-04-01 16:00:00  1004       ■■■■■■■■■■
2023-04-01 16:30:00  666        ■■■■■■■
2023-04-01 17:00:00  446        ■■■■■
2023-04-01 17:30:00  276        ■■■
2023-04-01 18:00:00  510        ■■■■■
2023-04-01 18:30:00  588        ■■■■■■
2023-04-01 19:00:00  404        ■■■■
2023-04-01 19:30:00  286        ■■■
2023-04-01 20:00:00  2256       ■■■■■■■■■■■■■■■■■■■■■■
2023-04-01 20:30:00  1620       ■■■■■■■■■■■■■■■■
2023-04-01 21:00:00  684        ■■■■■■■
2023-04-01 21:30:00  312        ■■■
2023-04-01 22:00:00  166        ■■
2023-04-01 22:30:00  148        ■■
2023-04-01 23:00:00  86         ■
2023-04-01 23:30:00  144        ■■
2023-04-02 00:00:00  88         ■"
`;

exports[`cli consumption command "maxpower" should display a graph 1`] = `
"
Date                 Valeur (VA) Graphique                     
2023-04-01 12:06:20  4638        ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■"
`;

exports[`cli production command "dailyprod" should display a graph 1`] = `
"
Date        Valeur (Wh) Graphique                     
2023-04-01  643         ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■"
`;

exports[`cli production command "loadcurveprod" should display a graph 1`] = `
"
Date                 Valeur (W) Graphique                     
2023-04-01 00:30:00  0          
2023-04-01 01:00:00  0          
2023-04-01 01:30:00  0          
2023-04-01 02:00:00  0          
2023-04-01 02:30:00  0          
2023-04-01 03:00:00  0          
2023-04-01 03:30:00  0          
2023-04-01 04:00:00  0          
2023-04-01 04:30:00  0          
2023-04-01 05:00:00  0          
2023-04-01 05:30:00  0          
2023-04-01 06:00:00  0          
2023-04-01 06:30:00  0          
2023-04-01 07:00:00  0          
2023-04-01 07:30:00  0          
2023-04-01 08:00:00  0          
2023-04-01 08:30:00  0          
2023-04-01 09:00:00  0          
2023-04-01 09:30:00  0          
2023-04-01 10:00:00  0          
2023-04-01 10:30:00  0          
2023-04-01 11:00:00  132        ■■■■■■■■■■■
2023-04-01 11:30:00  112        ■■■■■■■■■■
2023-04-01 12:00:00  368        ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
2023-04-01 12:30:00  288        ■■■■■■■■■■■■■■■■■■■■■■■■
2023-04-01 13:00:00  16         ■■
2023-04-01 13:30:00  122        ■■■■■■■■■■
2023-04-01 14:00:00  102        ■■■■■■■■■
2023-04-01 14:30:00  26         ■■■
2023-04-01 15:00:00  24         ■■
2023-04-01 15:30:00  0          
2023-04-01 16:00:00  0          
2023-04-01 16:30:00  76         ■■■■■■■
2023-04-01 17:00:00  16         ■■
2023-04-01 17:30:00  0          
2023-04-01 18:00:00  0          
2023-04-01 18:30:00  4          ■
2023-04-01 19:00:00  0          
2023-04-01 19:30:00  0          
2023-04-01 20:00:00  0          
2023-04-01 20:30:00  0          
2023-04-01 21:00:00  0          
2023-04-01 21:30:00  0          
2023-04-01 22:00:00  0          
2023-04-01 22:30:00  0          
2023-04-01 23:00:00  0          
2023-04-01 23:30:00  0          
2023-04-02 00:00:00  0          "
`;
